include/mrdox/Plugin.hpp

```diff
@@ -0,0 +1,45 @@
+//
+// This is a derivative work. originally part of the LLVM Project.
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+#ifndef MRDOX_PLUGIN_HPP
+#define MRDOX_PLUGIN_HPP
+
+#include <mrdox/Platform.hpp>
+#include <mrdox/Support/Error.hpp>
+
+namespace clang {
+namespace mrdox {
+
+class Generator;
+class PluginEnvironment
+{
+public:
+    virtual
+    void
+    reportError(Error error) = 0;
+    virtual
+    void
+    addGenerator(std::unique_ptr<Generator> generator) = 0;
+};
+
+}
+}
+
+// the version gets delivers as an argument, because env might incompatible.
+// this function must return false on version conflicts!
+extern "C"
+MRDOX_SYMBOL_EXPORT
+bool
```

Comment:
Why does it have to return false?

Reply:
To indicate a version mismatch. If the version doesn't match I can't use any of the C++ APIs, but need to rely on a super primitive C API.

Reply:
But what does the caller of `MrDoxMain` do with the information? The tool doesn't call into the plugin, the plugin calls into the tool. So it is the plugin's job to know if the version is safe to call into. If the version is not safe then the plugin simply avoids installing itself. I dont understand how the bool is used.

Reply:
So we can report that a plugin couldn't load for this reason, because it can't use `report:error`.